lib/Dialect/Comb/CombFolds.cpp

```diff
@@ -988,6 +988,188 @@ OpFoldResult ICmpOp::fold(ArrayRef<Attribute> constants) {
   return {};
 }
 
+// Given a range of operands, computes the number of matching prefix and
+// suffix elements. This does not perform cross-element matching.
+static std::pair<size_t, size_t>
+computeCommonPrefixAndSuffix(const OperandRange &a, const OperandRange &b) {
+  size_t commonPrefixLength = 0;
+  size_t commonSuffixLength = 0;
+  size_t sizeA = a.size();
+  size_t sizeB = b.size();
+
+  for (; commonPrefixLength < std::min(sizeA, sizeB); commonPrefixLength++) {
+    if (a[commonPrefixLength] != b[commonPrefixLength]) {
+      break;
+    }
+  }
+
+  for (; commonSuffixLength < std::min(sizeA, sizeB) - commonPrefixLength;
+       commonSuffixLength++) {
+    if (a[sizeA - commonSuffixLength - 1] !=
+        b[sizeB - commonSuffixLength - 1]) {
+      break;
+    }
+  }
+
+  return {commonPrefixLength, commonSuffixLength};
+}
+
+static size_t getTotalWidth(const OperandRange &range) {
+  size_t totalWidth = 0;
+  for (auto operand : range) {
+    // getIntOrFloatBitWidth should never raise, since all arguments to ConcatOp
+    // are integers.
+    ssize_t width = operand.getType().getIntOrFloatBitWidth();
+    assert(width >= 0);
+    totalWidth += width;
+  }
+  return totalWidth;
+}
+
+static Optional<Value> findFirstNonEmptyValue(const OperandRange &range) {
```

Comment:
Oh wow, nice corner case.  I'd recommend handling this a slightly different way:

1) Canonicalize cat in general to remove zero bit operands.
2) Make your new `icmp` optimization bail out with no change if it runs into the first operand being a zero bit value.

This should make it safe, catch the same cases, and will be simpler.

Reply:
It turns out that Comb.td already enforce that ConcatOp cannot take in zero-width arguments, nor can it take in zero-length varidics args. I added some defensive checks, and gotten rid of `findFirstNonEmptyValue`